src/wui/buildingwindow.cc

```diff
@@ -510,27 +522,39 @@ void BuildingWindow::act_enhance(Widelands::DescriptionIndex id, bool csite) {
 		upcast(Widelands::ConstructionSite, construction_site, building);
 		assert(construction_site);
 		if (SDL_GetModState() & KMOD_CTRL) {
-			igbase()->game().send_player_enhance_building(
-			   *construction_site, construction_site->building().enhancement(), false);
-		} else {
-			show_enhance_confirm(dynamic_cast<InteractivePlayer&>(*igbase()), *construction_site,
-			                     construction_site->get_info().becomes->enhancement(), true);
+			if (Widelands::Game* game = ibase()->get_game()) {
+				game->send_player_enhance_building(
+				   *construction_site, construction_site->building().enhancement(), false);
+			} else {
+				NEVER_HERE();  // TODO(Nordfriese / Scenario Editor): implement
+			}
+		} else if (upcast(InteractivePlayer, ipl, ibase())) {
+			show_enhance_confirm(
+			   *ipl, *construction_site, construction_site->get_info().becomes->enhancement(), true);
 		}
 		return;
 	}
 
 	if (SDL_GetModState() & KMOD_CTRL) {
 		if (building->get_playercaps() & Widelands::Building::PCap_Enhancable) {
-			igbase()->game().send_player_enhance_building(*building, id, false);
+			if (Widelands::Game* game = ibase()->get_game()) {
```

Comment:
We should do this dynamic cast once when constructing the BuildingWindow and save it to a variable. Then we only need to check against nullptr. Same for the interactive player.

Reply:
In #4297 I'm introducing a new `virtual bool EditorGameBase::is_game()`. Perhaps when that is in we could change that function here to 
```
Widelands::Game* InteractiveBase::get_game() const {
	return egbase_.is_game() ? &static_cast<Widelands::Game&>(egbase_) : nullptr;
}
````

Reply:
You still need to get the object though, so that still means dynamic casts. So, we should benefit from doing that dynamic cast only once in the constructor and sticking it into a protected variable. Of course, the constructor can then use is_game() once the other branch is in :)